<a href="https://colab.research.google.com/github/LuisaBeccar/proyecto_nlp/blob/main/classification_apartments.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

Mi csv creado con chetgpt, prompt:
"Quiero desarrollar un algoritmo de machine learning de clasificacion de texto. Para eso quiero que me crees una lista de 500 comentarios distintos de no mas de 100 palabras cada uno, que describan departamentos. la mitad que describan departamentos que tengan balcón y la otra mitad no. Dame la respuesta en formato de csv, con dos columnas, la primera que sea el comentario, y la segunda que diga 1 si tiene balcon y 0 si no tiene balcon." <br>
"mirando el archivo, hay 4 comentarios distintos cada uno se repite 125 veces. No cumple con mi condicion de que cada comentario sea unico. Por favor corrige ese requerimineto y dame un csv con 500 comentarios unicos, distintos dentre si. 250 que incluyan en su descripcion la presencia de espacio exterior tipo balcon, usando sinonimos del mismo, y los otros 250 que no lo mencionen o que nieguen que posean balcon o sus sinonimos. El fomato de csv con la priemra columna llamada "comentario" y la  segunda "balcon", donde si tiene balcón el valor sea 1 y si no lo tiene, sea 0." <br>
"Eso está mejor, ahora veo 16 comentarios diferentes, pero cada uno se repite varias veces... quiero que haya al menos 50 comentarios diferentes, unicos con las características que te describi arriba. Por favor construye más comentarios unicos distintos entre si para cumplir los objetivos 500 comentarios unico y distintos, de menos de 100 palabras sobre departamentos, para clasificar con o sin balcon"

In [3]:
df = pd.read_csv("https://raw.githubusercontent.com/LuisaBeccar/proyecto_nlp/refs/heads/main/comentarios_departamentos.csv")
df

comentario  balcon
0    Hermoso departamento con amplio balcón privado...       1
1    Apartamento con terraza perfecta para disfruta...       1
2    Departamento con balcón que ofrece una vista p...       1
3    Espacioso departamento con acceso a un balcón ...       1
4    Lindo apartamento con balcón ideal para desayu...       1
..                                                 ...     ...
96   Apartamento sin terraza, pero con una cochera ...       0
97   Departamento sin balcón, pero con un diseño vi...       0
98   Apartamento sin espacio exterior, pero con acc...       0
99   Departamento sin balcón, pero con ventanas de ...       0
100  Apartamento sin terraza, pero con conexión ráp...       0

[101 rows x 2 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101 entries, 0 to 100
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   comentario  101 non-null    object
 1   balcon      101 non-null    int64 
dtypes: int64(1), object(1)
memory usage: 1.7+ KB


In [5]:
df.comentario.value_counts()

comentario
Hermoso departamento con amplio balcón privado con vista al parque.               1
Apartamento sin terraza, pero con electrodomésticos incluidos.                    1
Apartamento sin espacio exterior, pero con una hermosa biblioteca interna.        1
Departamento sin balcón, pero con estacionamiento privado incluido.               1
Apartamento sin terraza, pero con transporte público a la puerta.                 1
                                                                                 ..
Apartamento con balcón semicubierto y espacio para niños.                         1
Luminoso departamento con balcón al este y sol matinal.                           1
Apartamento con balcón integrado al living, con puertas corredizas.               1
Departamento con terraza y parrilla de acero inoxidable.                          1
Apartamento sin terraza, pero con conexión rápida a internet por fibra óptica.    1
Name: count, Length: 101, dtype: int64

In [7]:
df.drop_duplicates(inplace=True)
df.comentario.value_counts()

comentario
Hermoso departamento con amplio balcón privado con vista al parque.               1
Apartamento sin terraza, pero con electrodomésticos incluidos.                    1
Apartamento sin espacio exterior, pero con una hermosa biblioteca interna.        1
Departamento sin balcón, pero con estacionamiento privado incluido.               1
Apartamento sin terraza, pero con transporte público a la puerta.                 1
                                                                                 ..
Apartamento con balcón semicubierto y espacio para niños.                         1
Luminoso departamento con balcón al este y sol matinal.                           1
Apartamento con balcón integrado al living, con puertas corredizas.               1
Departamento con terraza y parrilla de acero inoxidable.                          1
Apartamento sin terraza, pero con conexión rápida a internet por fibra óptica.    1
Name: count, Length: 101, dtype: int64

In [8]:
df.balcon.value_counts()

balcon
1    51
0    50
Name: count, dtype: int64

Bueno, he conseguido crear esta lista de 101 comentarios: 51 clasificados como con balcon y 50 sin balcon. Vamos a usarlo para lo que nos compete: crear un modelo de clasificacion de texto que prediga si el depto tiene o no balcon en base al comentario.

In [9]:
from sklearn.feature_extraction.text import CountVectorizer


# Initialize the Count Vectorizer
count_vectorizer = CountVectorizer()

# Fit the model and transform the data
count_matrix = count_vectorizer.fit_transform(df.comentario)

In [10]:
count_matrix

<101x250 sparse matrix of type '<class 'numpy.int64'>'
	with 872 stored elements in Compressed Sparse Row format>


acá vemos que la matriz generada tiene las 101 filas de comentarios, y tokens disitntos 250

In [11]:
count_matrix.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 1, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 1, 0]])

In [12]:
# Convert the count matrix to a DataFrame for better readability
count_df = pd.DataFrame(count_matrix.toarray(), columns=count_vectorizer.get_feature_names_out())

# Display the DataFrame
count_df


24  acceso  acero  acogedor  acogedores  acondicionado  aire  \
0     0       0      0         0           0              0     0   
1     0       0      0         0           0              0     1   
2     0       0      0         0           0              0     0   
3     0       1      0         0           0              0     0   
4     0       0      0         0           0              0     1   
..   ..     ...    ...       ...         ...            ...   ...   
96    0       0      0         0           0              0     0   
97    0       0      0         0           0              0     0   
98    0       1      0         0           0              0     0   
99    0       0      0         0           0              0     0   
100   0       0      0         0           0              0     0   

     aislamiento  aislantes  al  ...  ventanas  verde  vertical  vidrio  \
0              0          0   1  ...         0      0         0       0   
1              0          0   0  ...         0      0         0       0   
2              0          0   0  ...         0      0         0       0   
3              0          0   0  ...         0      0         0       0   
4              0          0   1  ...         0      0         0       0   
..           ...        ...  ..  ...       ...    ...       ...     ...   
96             0          0   0  ...         0      0         0       0   
97             0          0   0  ...         0      0         0       0   
98             0          0   0  ...         0      0         0       0   
99             1          0   0  ...         1      0         0       0   
100            0          0   0  ...         0      0         0       0   

     vintage  vista  vistas  zona  óptica  último  
0          0      1       0     0       0       0  
1          0      0       0     0       0       0  
2          0      1       0     0       0       0  
3          0      0       0     0       0       0  
4          0      0       0     0       0       0  
..       ...    ...     ...   ...     ...     ...  
96         0      0       0     0       0       0  
97         1      0       0     0       0       0  
98         0      0       0     0       0       0  
99         0      0       0     0       0       0  
100        0      0       0     0       1       0  

[101 rows x 250 columns]

In [13]:
count_df.shape

(101, 250)

# ahora hacer la matriz pero sacando las stopwords , spanish

In [14]:
%%capture
!pip install nltk
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords


In [15]:

# Initializar  Count Vectorizer con Spanish stop words
spanish_stop_words = stopwords.words('spanish')
count_vectorizer = CountVectorizer(stop_words=spanish_stop_words)


# Fit the model and transform the data
count_matrix = count_vectorizer.fit_transform(df.comentario)



# Convert the count matrix to a DataFrame for better readability
count_df = pd.DataFrame(count_matrix.toarray(), columns=count_vectorizer.get_feature_names_out())

# Display the DataFrame
count_df


24  acceso  acero  acogedor  acogedores  acondicionado  aire  \
0     0       0      0         0           0              0     0   
1     0       0      0         0           0              0     1   
2     0       0      0         0           0              0     0   
3     0       1      0         0           0              0     0   
4     0       0      0         0           0              0     1   
..   ..     ...    ...       ...         ...            ...   ...   
96    0       0      0         0           0              0     0   
97    0       0      0         0           0              0     0   
98    0       1      0         0           0              0     0   
99    0       0      0         0           0              0     0   
100   0       0      0         0           0              0     0   

     aislamiento  aislantes  alargado  ...  ventanas  verde  vertical  vidrio  \
0              0          0         0  ...         0      0         0       0   
1              0          0         0  ...         0      0         0       0   
2              0          0         0  ...         0      0         0       0   
3              0          0         0  ...         0      0         0       0   
4              0          0         0  ...         0      0         0       0   
..           ...        ...       ...  ...       ...    ...       ...     ...   
96             0          0         0  ...         0      0         0       0   
97             0          0         0  ...         0      0         0       0   
98             0          0         0  ...         0      0         0       0   
99             1          0         0  ...         1      0         0       0   
100            0          0         0  ...         0      0         0       0   

     vintage  vista  vistas  zona  óptica  último  
0          0      1       0     0       0       0  
1          0      0       0     0       0       0  
2          0      1       0     0       0       0  
3          0      0       0     0       0       0  
4          0      0       0     0       0       0  
..       ...    ...     ...   ...     ...     ...  
96         0      0       0     0       0       0  
97         1      0       0     0       0       0  
98         0      0       0     0       0       0  
99         0      0       0     0       0       0  
100        0      0       0     0       1       0  

[101 rows x 226 columns]

In [16]:
count_df.shape

(101, 226)

Ac vemos como de 250 tokens pasamos a 226... pequeña reduccion de dimensionalidad

In [17]:
X = count_matrix.toarray()
X

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 1, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 1, 0]])

In [18]:
y = df.balcon
y

0      1
1      1
2      1
3      1
4      1
      ..
96     0
97     0
98     0
99     0
100    0
Name: balcon, Length: 101, dtype: int64

In [19]:
from sklearn.linear_model import LogisticRegression

In [20]:
clf = LogisticRegression().fit(X, y)
clf

LogisticRegression()

In [31]:
txt_test = input("Escriba un comentario de menos de 100 palabras sobre un departamento:")

Escriba un comentario de menos de 100 palabras sobre un departamento:terraza de seis metros cuadrados con vista al parque lezama


 para poder clasificar mi texto tengo que transformarlo en 1 y 0.. solo con transform (antes en el texto de training era fit.transform, para entrenar el modelo)... y a array. Lo guardo en variable x_test


In [32]:
x_test = count_vectorizer.transform([txt_test])

Finalemnte si usar el clasificador con el array de mi texto

In [33]:
clf.predict(x_test)

array([1])

Disclaimer: probando el codigo, veo que si escribo algo con "balcon" sin tilde en la "o", no lo reconoce como balcón.

# Text Classification

mi dataset es bastante chico: 101 filas, asique separarlo en train y test seria achicar mas la muestra de training... pero si lo hiciera, seria algo asi:

In [34]:
df2 = pd.read_csv("https://raw.githubusercontent.com/LuisaBeccar/proyecto_nlp/refs/heads/main/comentarios_departamentos.csv")

In [35]:
X = df2.comentario
y = df2.balcon

In [36]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [ ]:
"""
%%capture
!pip install nltk
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
"""

In [37]:
# Initialize the Count Vectorizer
spanish_stop_words = stopwords.words('spanish')
count_vectorizer2 = CountVectorizer(stop_words=spanish_stop_words)

# Fit the model and transform the data
X_train_v = count_vectorizer2.fit_transform(X_train)


clf = LogisticRegression()
clf.fit(X_train_v, y_train)


LogisticRegression()

In [38]:
X_test.shape

(26,)

In [42]:
X_test_v = count_vectorizer2.transform(X_test)

In [44]:
y_pred = clf.predict(X_test_v)

y se podria calcular la accuracy del modelo

In [45]:
from sklearn.metrics import accuracy_score

In [46]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.4f}")

Accuracy: 0.5385


mediocre, posiblemente por la redicida cantida de casos de entrenamiento

#TF-IDF

In [50]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer(stop_words=spanish_stop_words)

# Transform the text data
X_train_tfidf = vectorizer.fit_transform(X_train)
X_test_tfidf = vectorizer.transform(X_test)

In [53]:
X_train_tfidf

<75x178 sparse matrix of type '<class 'numpy.float64'>'
	with 402 stored elements in Compressed Sparse Row format>

In [55]:
X_train_tfidf.toarray()

array([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.37877846, 0.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [56]:
clf3 = LogisticRegression()
clf3.fit(X_train_tfidf, y_train)


LogisticRegression()

In [57]:


# Make predictions
y_pred3 = clf3.predict(X_test_tfidf)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred3)
print(f"Accuracy: {accuracy:.4f}")

Accuracy: 0.5000
